### PR TITLE
curve read-only reentrancy

### PR DESCRIPTION
**Pull request by joaquinlpereyra**
Worked on Curve read-only re-entrancy problems with a focus on the attack on Qi specifically. 

- I wrote the meat of the README, missing sources and graphs. Sources should mention issue #43 which brought this to our attention
- I have a semi-finished implementation of the attack, with the reentrancy to the evil contract working. Missing asking for a borrow with the mispriced LP-tokens.

The (trace transaction)[https://tx.eth.samczsun.com/polygon/0xb8efe839da0c89daa763f39f30577dc21937ae351c6f99336a0017e63d387558] is very big but should be helpful if anyone wants to take a look.

### CODE REVIEW

**Comment by nine-december** [comment]
Worked up to the repayment of one token. A swap and the repayment of the other one remaining to close the loan and finish the attack.

File: test/Reentrancy/CurvePoolOracle/QiReadOnlyReentrancy.attack.sol [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        uint256[] memory ,
+        bytes memory
+    ) external payable {
+        // Sensible requires to aid development
+        require(msg.sender == address(balancer), "only callable by balancer");
+        require(tokens.length == 2 && tokens.length == amounts.length, "length missmatch");
+        require(address(tokens[0]) == address(WMATIC));
+        require(address(tokens[1]) == address(stLIDOMATIC));
+
+        console.log('===== 3. HANDLE BALANCER FLASHLOAN ======');
```
Comment (nine-december) [inline test/Reentrancy/CurvePoolOracle/QiReadOnlyReentrancy.attack.sol:242]: It will revert here because no flashloan has been repaid yet.

**Comment by nine-december** [comment]
Ready to goooo 🎉

**Review by joaquinlpereyra — COMMENTED** [review]
Fixed a slight mistake in the README, otherwise nice to finally have this ready 👏🏿... ah I can't approve because I corrected something myself :)

**Review by juli — APPROVED** [review]

**Pull request merged**